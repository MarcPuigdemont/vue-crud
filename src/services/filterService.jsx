const contains = (string, substring) => string.toLowerCase().indexOf(substring.toLowerCase()) > -1;

const filterByStringAndServices = (airlines, string, serviceList = []) => {
  const matchingAirlines = string.trim() === '' ? airlines : airlines.filter(a => contains(a.iata, string) || contains(a.name, string));

  const numOfServices = serviceList.length;
  return matchingAirlines.filter(airline => serviceList.filter(s => airline.services[s]).length === numOfServices);
};

// eslint-disable-next-line import/prefer-default-export
export { filterByStringAndServices };
