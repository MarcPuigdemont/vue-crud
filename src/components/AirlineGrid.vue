<template>
  <b-row class="w-100">
    <b-col v-for="(airline, index) in filteredAirlines" :key="index" class="pr-2 pb-2 col-lg-3 col-md-3 col-sm-4 col-xs-6">
      <AirlineGridItem :airline="airline" @edit="edit" />
    </b-col>
  </b-row>
</template>
<script>
import { BCol, BRow } from 'bootstrap-vue';
import AirlineGridItem from './AirlineGridItem.vue';
import { filterByStringAndServices } from '@/services/filterService.jsx';

export default {
  name: 'AirlineGrid',
  components: {
    AirlineGridItem,
    BCol,
    BRow,
  },
  props: {
    airlines: Array,
    filter: String,
    serviceFilter: Array,
  },
  computed: {
    filteredAirlines() {
      return filterByStringAndServices(this.airlines, this.filter, this.serviceFilter);
    },
  },
  methods: {
    edit(airline) {
      this.$emit('edit', airline);
    },
  },
};
</script>
