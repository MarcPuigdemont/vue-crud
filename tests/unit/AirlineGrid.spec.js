import { mount, shallowMount } from '@vue/test-utils';
import AirlineGrid from '@/components/AirlineGrid.vue';
import * as filterService from '@/services/filterService.jsx';

const propsData = {
  airlines: [
    {
      iata: 'BX',
      name: 'Airline name',
      primary_color: '#ffffff',
      secondary_color: '#000000',
      services: { bags: true, checkin: false, seats: true },
    },
    {
      iata: 'AX',
      name: 'Airline name 2',
      primary_color: '#ffaaff',
      secondary_color: '#00aa00',
      services: { bags: false, checkin: true, seats: true },
    },
  ],
  filter: '',
  serviceFilter: [],
};

describe('AirlineGrid', () => {
  test('should match the snapshot', () => {
    const wrapper = mount(AirlineGrid, {
      propsData,
      stubs: {
        AirlineGridItem: '<div class="airline-grid-item"></div>',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  test('should emit the same airline as it gets from edit signal', () => {
    filterService.filterByStringAndServices = jest.fn();
    const wrapper = shallowMount(AirlineGrid, {
      propsData,
    });

    wrapper.vm.edit(propsData.airlines[0]);
    expect(wrapper.emitted('edit')[0][0]).toEqual(propsData.airlines[0]);
    expect(filterService.filterByStringAndServices).toBeCalledWith(propsData.airlines, propsData.filter, propsData.serviceFilter);
  });
});
