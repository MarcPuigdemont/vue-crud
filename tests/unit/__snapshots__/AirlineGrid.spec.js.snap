// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirlineGrid should match the snapshot 1`] = `
<div
  class="row w-100"
>
  <div
    class="pr-2 pb-2 col-lg-3 col-md-3 col-sm-4 col-xs-6 col"
  >
    <div
      class="airline-grid-item"
    />
  </div>
  <div
    class="pr-2 pb-2 col-lg-3 col-md-3 col-sm-4 col-xs-6 col"
  >
    <div
      class="airline-grid-item"
    />
  </div>
</div>
`;
